Build error involving wrong importmap passed to vendored external tests

Originally reported by @ixdy. To reproduce, checkout https://github.com/ixdy/kubernetes at 8704a68a3712d35e778043d6e4109913449677e9 and run this command:

```
$ bazel build staging/src/k8s.io/apimachinery/pkg/runtime:go_default_test
...
/usr/local/google/home/jayconrod/.cache/bazel/_bazel_jayconrod/4a2bdc82526db3ea358ea56d8f9e540e/sandbox/linux-sandbox/107/execroot/__main__/staging/src/k8s.io/apimachinery/pkg/runtime/conversion_test.go:34:94: cannot use "k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/runtime/testing".InternalComplex literal (type *"k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/runtime/testing".InternalComplex) as type "k8s.io/apimachinery/pkg/runtime".Object in argument to scheme.AddKnownTypeWithName:
	*"k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/runtime/testing".InternalComplex does not implement "k8s.io/apimachinery/pkg/runtime".Object (wrong type for DeepCopyObject method)
		have DeepCopyObject() "k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/runtime".Object
		want DeepCopyObject() "k8s.io/apimachinery/pkg/runtime".Object
...
```

--- a/go/private/context.bzl
+++ b/go/private/context.bzl
@@ -87,8 +87,8 @@ def _new_args(go):
 def _new_library(go, name = None, importpath = None, resolver = None, importable = True, testfilter = None, **kwargs):
     if not importpath:
         importpath = go.importpath
-    importmap = getattr(go._ctx.attr, "importmap", "")
-    if not importmap:
+        importmap = go.importmap
+    else:
         importmap = importpath
     pathtype = go.pathtype
     if not importable and pathtype == EXPLICIT_PATH:
@@ -160,31 +160,36 @@ def _infer_importpath(ctx):
     DEFAULT_LIB = "go_default_library"
     VENDOR_PREFIX = "/vendor/"
 
-    # Check if import path was explicitly set
-    path = getattr(ctx.attr, "importpath", "")
-
-    # are we in forced infer mode?
-    if path != "":
-        return path, EXPLICIT_PATH
-
-    # See if we can collect importpath from embeded libraries
-    # This is the path that fixes tests as well
+    # Check if paths were explicitly set, either in this rule or in an
+    # embedded rule.
+    attr_importpath = getattr(ctx.attr, "importpath", "")
+    attr_importmap = getattr(ctx.attr, "importmap", "")
+    embed_importpath = ""
+    embed_importmap = ""
     for embed in getattr(ctx.attr, "embed", []):
         if GoLibrary not in embed:
             continue
-        if embed[GoLibrary].pathtype == EXPLICIT_PATH:
-            return embed[GoLibrary].importpath, EXPLICIT_PATH
+        lib = embed[GoLibrary]
+        if lib.pathtype == EXPLICIT_PATH:
+            embed_importpath = lib.importpath
+            embed_importmap = lib.importmap
+            break
+
+    importpath = attr_importpath or embed_importpath
+    importmap = attr_importmap or embed_importmap or importpath
+    if importpath:
+        return importpath, importmap, EXPLICIT_PATH
 
     # Guess an import path based on the directory structure
     # This should only really be relied on for binaries
-    path = ctx.label.package
-    if ctx.label.name != DEFAULT_LIB and not path.endswith(ctx.label.name):
-        path += "/" + ctx.label.name
-    if path.rfind(VENDOR_PREFIX) != -1:
-        path = path[len(VENDOR_PREFIX) + path.rfind(VENDOR_PREFIX):]
-    if path.startswith("/"):
-        path = path[1:]
-    return path, INFERRED_PATH
+    importpath = ctx.label.package
+    if ctx.label.name != DEFAULT_LIB and not importpath.endswith(ctx.label.name):
+        importpath += "/" + ctx.label.name
+    if importpath.rfind(VENDOR_PREFIX) != -1:
+        importpath = importpath[len(VENDOR_PREFIX) + importpath.rfind(VENDOR_PREFIX):]
+    if importpath.startswith("/"):
+        importpath = importpath[1:]
+    return importpath, importpath, INFERRED_PATH
 
 def _get_go_binary(context_data):
     for f in context_data.sdk_tools:
@@ -238,7 +243,7 @@ def go_context(ctx, attr = None):
         "PATH": context_data.cgo_tools.compiler_path,
     })
 
-    importpath, pathtype = _infer_importpath(ctx)
+    importpath, importmap, pathtype = _infer_importpath(ctx)
     return GoContext(
         # Fields
         toolchain = toolchain,
@@ -255,6 +260,7 @@ def go_context(ctx, attr = None):
         crosstool = context_data.crosstool,
         package_list = context_data.package_list,
         importpath = importpath,
+        importmap = importmap,
         pathtype = pathtype,
         cgo_tools = context_data.cgo_tools,
         builders = builders,

--- a/tests/core/go_test/BUILD.bazel
+++ b/tests/core/go_test/BUILD.bazel
@@ -75,3 +75,25 @@ go_test(
     size = "small",
     srcs = ["only_testmain_test.go"],
 )
+
+go_test(
+    name = "external_importmap_test",
+    size = "small",
+    srcs = ["external_importmap_test.go"],
+    embed = [":external_importmap_lib"],
+    deps = [":external_importmap_dep"],
+)
+
+go_library(
+    name = "external_importmap_lib",
+    srcs = ["external_importmap_lib.go"],
+    importmap = "x/github.com/bazelbuild/rules_go/tests/core/go_test/external_importmap",
+    importpath = "github.com/bazelbuild/rules_go/tests/core/go_test/external_importmap",
+)
+
+go_library(
+    name = "external_importmap_dep",
+    srcs = ["external_importmap_dep.go"],
+    importpath = "github.com/bazelbuild/rules_go/tests/core/go_test/external_importmap_dep",
+    deps = [":external_importmap_lib"],
+)

--- a/tests/core/go_test/README.rst
+++ b/tests/core/go_test/README.rst
@@ -45,3 +45,14 @@ only_testmain_test
 ------------------
 Test that an `go_test`_ that contains a ``TestMain`` function but no tests
 still builds and passes.
+
+external_importmap_test
+----------------------
+
+Test that an external test package in `go_test`_ is compiled with the correct
+``importmap`` for the library under test. This is verified by defining an
+interface in the library under test and implementing it in a separate
+dependency.
+
+Verifies #1538.
+

--- a/tests/core/go_test/external_importmap_dep.go
+++ b/tests/core/go_test/external_importmap_dep.go
@@ -1,0 +1,11 @@
+package external_importmap_dep
+
+import (
+	"github.com/bazelbuild/rules_go/tests/core/go_test/external_importmap"
+)
+
+type Impl struct{}
+
+func (_ *Impl) DeepCopyObject() external_importmap.Object {
+	return nil
+}

--- a/tests/core/go_test/external_importmap_lib.go
+++ b/tests/core/go_test/external_importmap_lib.go
@@ -1,0 +1,5 @@
+package external_importmap
+
+type Object interface {
+	DeepCopyObject() Object
+}

--- a/tests/core/go_test/external_importmap_test.go
+++ b/tests/core/go_test/external_importmap_test.go
@@ -1,0 +1,10 @@
+package external_importmap_test
+
+import (
+	"github.com/bazelbuild/rules_go/tests/core/go_test/external_importmap"
+	"github.com/bazelbuild/rules_go/tests/core/go_test/external_importmap_dep"
+)
+
+var _ external_importmap.Object = &external_importmap_dep.Impl{}
+
+// Test passes if it compiles.

